File: langs/typescript/src/dialogs/nullDialogService.ts
/**
 * NullDialogService — null-object implementation of {@link IDialogService}.
 *
 * All methods return the safest default:
 * - `pickFileToOpen` / `pickFileToSave` → `null` (treat as cancel)
 * - `confirm` → `false` (avoids triggering destructive operations)
 * - `notify` → no-op
 *
 * Stateless — safe to share via {@link NullDialogService.INSTANCE}.
 * See spec/19-dialogs.md §4 and ADR-0017.
 */
import type { FileFilter, IModalDialogService, NotificationSeverity } from "./dialogService.js";
import type { ModalVM } from "./modalVM.js";

export class NullDialogService implements IModalDialogService {
  /** Shared singleton instance (the service holds no state). */
  static readonly INSTANCE: NullDialogService = new NullDialogService();

  // Singleton: consume via INSTANCE (matches the other null variants and C#).
  private constructor() {}

  pickFileToOpen(
    _filter?: FileFilter | null,
    _title?: string | null,
  ): Promise<string | null> {
    return Promise.resolve(null);
  }

  pickFileToSave(
    _filter?: FileFilter | null,
    _title?: string | null,
    _suggestedName?: string | null,
  ): Promise<string | null> {
    return Promise.resolve(null);
  }

  confirm(_message: string, _title?: string | null): Promise<boolean> {
    return Promise.resolve(false);
  }

  notify(
    _message: string,
    _title?: string | null,
    _severity?: NotificationSeverity,
  ): Promise<void> {
    return Promise.resolve();
  }

  present<T>(modalVm: ModalVM<T>): Promise<T> {
    modalVm.dismiss(modalVm.cancellationResult);
    return Promise.resolve(modalVm.cancellationResult);
  }
}

File: langs/swift/Sources/VMx/Dialogs/NullDialogService.swift
//
// NullDialogService — null-object variant of `DialogService`.
//
// See spec/19-dialogs.md §"Null variant" and ADR-0029.
//
// All operations complete with the safest possible default:
//   - pickFileToOpen / pickFileToSave → nil  (no file selected)
//   - confirm                         → false (not confirmed)
//   - notify                          → no-op (acknowledged silently)
//

/// Null-object implementation of `DialogService`. Suitable as a default
/// injection target when no real host UI is available (tests, previews, CLI).
public final class NullDialogService: DialogService {
    /// Shared singleton instance. The service is stateless.
    public static let INSTANCE = NullDialogService()

    public init() {}

    /// Returns `nil` — no file was selected (cancel default).
    public func pickFileToOpen(filter: FileFilter?, title: String?) async -> String? {
        nil
    }

    /// Returns `nil` — no file was selected (cancel default).
    public func pickFileToSave(filter: FileFilter?, title: String?, suggestedName: String?) async -> String? {
        nil
    }

    /// Returns `false` — not confirmed (safest default per DIA-007).
    public func confirm(_ message: String, title: String?) async -> Bool {
        false
    }

    /// No-op — notification is silently acknowledged.
    public func notify(_ message: String, title: String?, severity: NotificationSeverity) async {
        // intentional no-op
    }
}
